nodejs/week3/meal-sharing/src/backend/api/meals.js

```diff
@@ -0,0 +1,151 @@
+  response.json(meal);
+})*/
+
+
+// GET api/meals/ query parameters and you should make sure that you can combine them
+/*router.get("/", async (request, response) => {
+    if (request.query.maxPrice) {
+      const maxPriceMeal = await knex("meal").select().where('price', '<', Number(request.query.maxPrice))
+      response.json(maxPriceMeal)
+    };
+    if (request.query.availableReservations) {
+      const availableMeals = await knex("meal")
+      .join('reservation', 'mealid', 'reservations.mealid')
+      .select('meal.id', 'title', 'maxreservatins')
+      .where("maxreservation", ">", "numberofguests")
+      response.json(availableMeals)
+    };
+    // a bit dif here just to try som new
+    if ('title' in request.query) {
+      const title = await knex("meal").where(meal.title);
+      const getTitle = meals.find((meal) =>
+      meal.title.toLowerCase().includes(title.toLowerCase()),
+      response.send(getTitle)
+      )};
+
+    if ("createdAfter" in req.query) {
+      const createdAfterMeals = await knex("meal").where("createddate", ">", req.query.createdAfter)
+      response.json(createdAfterMeals)
+    };
+
+    if (limit.request.query) {
+      const limitMeals =  await knex("meal").where.()
+      if (meals.length >= 6) {
+        meals.splice(0, limit)
+      }
+          response.send(limit)
+  });*/
+
+  router.get("/", async (request, response) => {
+    const meals = await knex("meal").select();
+    let filteredMealsByPrice = [];
+    let filteredMealsByTitle = [];
+    let filteredByDate = [];
+    let filteredByLimit = [];
+    let filteredByReservation = [];
+    if (request.query.maxPrice) {
+      filteredMealsByPrice = meals.filter(meal => Number(meal.price) < Number(request.query.maxPrice))
+    } else { 
+      filteredMealsByPrice = meals
+    }
+    if (request.query.title) {
+      filteredMealsByTitle = filteredMealsByPrice.filter(meal =>
+      meal.title.toLowerCase().includes(request.query.title.toLowerCase()))
+    } else {
+      filteredMealsByTitle = filteredMealsByPrice
+    }
+    if (request.query.createdAfter) {
+      filteredByDate = filteredMealsByTitle.filter(meal => 
+        meal.createddate > new Date(request.query.createdAfter))
+    } else {
+      filteredByDate = filteredMealsByTitle
+    }
+    if (request.query.limit) {
+      filteredByLimit = filteredByDate.splice(0, request.query.limit)
+    } else {
+      filteredByLimit = filteredByDate
+    }
+    /*for(let i=0; i < reservation.mealid.length; i++) 
+         const reservationId = reservation.mealid;
+         if(request.query.availableReservations) {
+          filteredByReservation = filteredByLimit.filter( meal => meal.id === reservationId)
+         } else {
+          filteredByReservation = filteredByLimit
+        }*/
+
+      /* for the last part -availableReservations I tried 2 solutions, the first was to use the .join knex method but that made
```

Comment:
So, you have the meals table, and each meal can have several reservations. This a one (meal) to-many (reservations) relationship.
One way to go with the availableReservations param is that you need to find out how many reservations a meal has, and then compare that with the max_reservations number.
I suggest you first create a query that does a left join with the reservations (this way you take into account meals that don't have even one reservation), group by and count the reservations.meal_id and use .having to return the meals ids with a higher max_reservations then the count. That way you find out how many reservations a meal has and compare it with the parameter to return the meals Ids with 'available reservations'. So, after you have the meals id you can just do a whereIn query.
I believe something close to this will do the job:
```
const availableMealsIds = knex('meal')
    .select('meal.id', 'meal.max_reservations')
    .leftJoin('reservation', { 'meal.id': 'reservation.meal_id' })
    .count({'reservations' : 'reservation.meal_id'})
    .groupBy('meal.id')
    .having(knex.raw('meal.max_reservations > reservations'));
```
And then, simply use these Ids to get the meals you need :)